src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
             if (string.IsNullOrEmpty(value))
                 return value;
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            int cSafe = 0;
+            int cSpaces = 0;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+                if (ch == ' ')
+                {
+                    cSpaces++;
+                }
+                else if (IsUrlSafeChar(ch))
+                {
+                    cSafe++;
+                }
+            }
+
+            if (cSpaces == 0 && cSafe == value.Length)
+            {
+                // Nothing to expand
+                return value;
+            }
+
+            int unsafeByteCount = Encoding.UTF8.GetByteCount(value) - cSafe - cSpaces;
+            int byteIndex = unsafeByteCount * 2;
+            
+            // For perf, we avoid allocating a new array for the encoded bytes, and
+            // instead encode in place
+            byte[] newBytes = new byte[cSafe + cSpaces + unsafeByteCount + byteIndex];
+            Encoding.UTF8.GetBytes(value, 0, value.Length, newBytes, byteIndex);
+            
+            GetEncodedBytes(newBytes, byteIndex, newBytes);
```

Comment:
I don't understand the logic here.  It seems like the goal is to use a single array.  But there's no guarantee that all of the characters that need URL encoding come at the end of the array... they could be interspersed throughout the entire array.